argoverse/evaluation/detection_utils.py

```diff
@@ -0,0 +1,175 @@
+    """
+    if method == InterpType.ALL:
+        prec_interp = np.maximum.accumulate(prec[::-1])[::-1]
+    else:
+        raise NotImplemented("This interpolation method is not implemented!")
+    return prec_interp
+
+
+def compute_match_matrix(dts: np.ndarray, gts: np.ndarray, metric: SimFnType) -> np.ndarray:
+    """Calculate the match matrix between detections and ground truth labels,
+    using a specified similarity function.
+
+    Args:
+        dts: Detections.
+        gts: Ground truth labels.
+        metric: Similarity metric type.
+
+    Returns:
+        Interpolated precision at all recall levels.
+    """
+    if metric == SimFnType.CENTER:
+        dt_centers = np.array([dt.translation for dt in dts])
+        gt_centers = np.array([gt.translation for gt in gts])
+        sims = -cdist(dt_centers, gt_centers)
+    else:
+        raise NotImplemented("This similarity metric is not implemented!")
+    return sims
+
+
+def dist_fn(dts: pd.DataFrame, gts: pd.DataFrame, metric: DistFnType) -> np.ndarray:
+    """Distance functions between detections and ground truth.
+
+    Args:
+        dts: Detections.
+        gts: Ground truth labels.
+        metric: Distance function type.
+
+    Returns:
+        Distance between the detections and ground truth, using the provided metric.
+    """
+    if metric == DistFnType.TRANSLATION:
+        dt_centers = np.vstack(dts["translation"].array)
+        gt_centers = np.vstack(gts["translation"].array)
+        trans_errors = np.linalg.norm(dt_centers - gt_centers, axis=1)
+        return trans_errors
+    elif metric == DistFnType.SCALE:
+        dt_dims = dts[["width", "length", "height"]]
+        gt_dims = gts[["width", "length", "height"]]
+        inter = np.minimum(dt_dims, gt_dims).prod(axis=1)
+        union = np.maximum(dt_dims, gt_dims).prod(axis=1)
+        scale_errors = 1 - (inter / union)
+        return scale_errors
+    elif metric == DistFnType.ORIENTATION:
+        # re-order quaternions to go from Argoverse format to scipy format, then the third euler angle (z) is yaw
+        dt_quats = np.vstack(dts["quaternion"].array)
+        dt_yaws = R.from_quat(quat_argo2scipy_vectorized(dt_quats)).as_euler("xyz")[:, 2]
+
+        gt_quats = np.vstack(gts["quaternion"].array)
+        gt_yaws = R.from_quat(quat_argo2scipy_vectorized(gt_quats)).as_euler("xyz")[:, 2]
+
+        signed_orientation_errors = normalize_angle(dt_yaws - gt_yaws)
+        orientation_errors = np.abs(signed_orientation_errors)
+        return orientation_errors
+    else:
+        raise NotImplemented("This distance metric is not implemented!")
+
+
+def normalize_angle(angle: float) -> float:
+    """Map angle (in radians) to [0, 2π).
+
+    Args:
+        angle: Angle (in radians).
+
+    Returns:
+        The angle (in radians) mapped to the interval [0, 2π).
+    """
+    period = 2 * np.pi
+    phase_shift = np.pi
+    return (angle + np.pi) % period - phase_shift
```

Comment:
Isn't it the same thing as https://github.com/argoai/argoverse-api/blob/master/argoverse/evaluation/eval_tracking.py#L85?

Reply:
Yes, this is an alternative formula. That works, but we need to add vectorization

Reply:
Cool. I would probably prefer to have that formula in one place (with vectorization), rather than in two places.

But wouldn't we want the detection error range to be only in [0,pi]? since we would take the smaller of the two possible angles (clockwise, or counter-clockwise)?

"Average Orientation Error (AOE) is the smallest yaw angle difference
between prediction and ground truth (radians)." https://arxiv.org/pdf/1903.11027.pdf

Reply:
yea, you're right. docstring typo. do you want me to just modify the other, original function?

Reply:
sure, that would be good to make the other one support vectorization. Thanks Ben

Reply:
@alliecc @seanremy @janesjanes what are your thoughts on this one? which would be more readable? and should we vectorize the one from the tracking eval script, and use it in both places?